crates/polars-core/src/chunked_array/ops/is_in.rs

```diff
@@ -1,44 +1,114 @@
 use std::hash::Hash;
+use std::ops::Add;
+
+use arrow::compute;
+use arrow::types::simd::Simd;
 
 use crate::prelude::*;
-use crate::utils::{try_get_supertype, CustomIterTools};
+#[cfg(feature = "search_sorted")]
+use crate::series::ops::search_sorted::*;
+use crate::series::IsSorted;
+use crate::utils::try_get_supertype;
 
 unsafe fn is_in_helper<T, P>(ca: &ChunkedArray<T>, other: &Series) -> PolarsResult<BooleanChunked>
 where
     T: PolarsNumericType,
-    P: Eq + Hash + Copy,
+    <T::Native as Simd>::Simd: Add<Output = <T::Native as Simd>::Simd>
+        + compute::aggregate::Sum<T::Native>
+        + compute::aggregate::SimdOrd<T::Native>,
+    P: Eq + Hash + Copy + Ord + Default,
 {
-    let mut set = PlHashSet::with_capacity(other.len());
+    let can_use_binary_search =
+        cfg!(feature = "search_sorted") && !matches!(other.is_sorted_flag(), IsSorted::Not);
+    let ca_name = ca.name();
 
     let other = ca.unpack_series_matching_type(other)?;
-    other.downcast_iter().for_each(|iter| {
-        iter.into_iter().for_each(|opt_val| {
-            // Safety
-            // bit sizes are/ should be equal
-            let ptr = &opt_val.copied() as *const Option<T::Native> as *const Option<P>;
-            let opt_val = *ptr;
-            set.insert(opt_val);
-        })
-    });
+    let other_has_nulls = other.null_count() > 0;
 
-    let name = ca.name();
-    let mut ca: BooleanChunked = ca
-        .into_iter()
-        .map(|opt_val| {
-            // Safety
-            // bit sizes are/ should be equal
-            let ptr = &opt_val as *const Option<T::Native> as *const Option<P>;
-            let opt_val = *ptr;
-            set.contains(&opt_val)
-        })
-        .collect_trusted();
-    ca.rename(name);
+    let mut ca: BooleanChunked = if !can_use_binary_search && ca.len() <= 10 {
+        // For small, unsorted `ca`, constructing the set of values in `other` is slower than just using many `.equal()` calls.
+        ca.into_iter()
+            .map(|opt_val| opt_val.map_or(other_has_nulls, |ca_val| other.equal(ca_val).any()))
+            .collect_trusted()
+    } else if can_use_binary_search && ca.len() * (other.len().ilog2() as usize) < other.len() {
+        // For sorted `other`, binary search can be faster than constructing the set of values in `other`.
+        // Complexity of hash set: construction: O(other) * O(1) + lookups: O(ca) * O(1)
+        // Complexity of binary search: lookups: O(ca) * O(log2(other))
+        // O(hash set)       <  O(binary search)              <=>
+        // O(other) + O(ca)  <  O(ca) * O(log2(other))        <=>
+        // O(other)          <  O(ca) * (O(log2(other)) - 1)  <=>
```

Comment:
O(other) <  O(ca) * (O(log2(other)) - 1) <=>  O(other) <  O(ca) * O(log2(other))
 
 Is this a typo?